### PR TITLE
Allow row selection in fixed tables

### PR DESCRIPTION
**Pull request by AdamWard1995**
### This project uses [semver](semver.org), please check the scope of this pr:

- [ ] #none# - documentation fixes and/or test additions
- [ ] #patch# - backwards-compatible bug fix
- [X] #minor# - adding functionality in a backwards-compatible manner
- [ ] #major# - incompatible API change

Add functionality for selecting table rows to **frost-fixed-table**. Needed for an internal story.

# CHANGELOG
 * **Added** functionality for selecting table rows in **frost-fixed-table**

### CODE REVIEW

**Comment by AdamWard1995** [comment]
Realized that I left changes in the demo I didn't want to, will remove those

File: addon/components/frost-fixed-table.js [hunk 1]
```diff
@@ -278,6 +329,32 @@ export default Component.extend({
...
+    const sides = [
+      {
+        header: this.get('headerLeftSelector'),
+        body: this.get('bodyLeftSelector')
+      }, {
+        header: this.get('headerRightSelector'),
+        body: this.get('bodyRightSelector')
+      }
+    ]
+    for (let side of sides) {
```
Comment (dafortin) [inline addon/components/frost-fixed-table.js]: I will recommend using `reduce` or `map` instead

File: addon/components/frost-fixed-table.js [hunk 2]
```diff
@@ -278,6 +329,32 @@ export default Component.extend({
...
+      const headerCells = this.$(`${side.header} .frost-table-header-cell`)
+      for (let pos = 0; pos < headerCells.length; ++pos) {
```
Comment (dafortin) [inline addon/components/frost-fixed-table.js]: Any reason to use `++pos` instead of `pos++` 

This change will avoid doing a `pos + 1`

File: addon/components/frost-fixed-table.js (hunk 1 above)
Comment (dafortin) [inline addon/components/frost-fixed-table.js]: or maybe a `forEach`

File: addon/components/frost-fixed-table.js [hunk 3]
```diff
@@ -335,13 +412,28 @@ export default Component.extend({
 
   // == Lifecycle Hooks =======================================================
 
+  init () {
+    this._super(...arguments)
+    const itemKey = this.get('itemKey')
```
Comment (dafortin) [inline addon/components/frost-fixed-table.js]: Is `itemKey` the name of an attribute?

File: addon/components/frost-fixed-table.js [hunk 4]
```diff
@@ -335,13 +412,28 @@ export default Component.extend({
...
+    if (itemKey) {
+      this.set('_itemComparator', function (lhs, rhs) {
```
Comment (dafortin) [inline addon/components/frost-fixed-table.js]: This is your default comparison function? Maybe we can isolate this in a util or something? or a function?

File: addon/components/frost-fixed-table.js [hunk 5]
```diff
@@ -335,13 +412,28 @@ export default Component.extend({
...
+  },
+
   /**
    * Set up synced scrolling as well as calculating padding for middle sections
    */
   didRender () {
     this._super(...arguments)
     this.setupBodyHeights()
     this.setupHoverProxy()
+    this.setupLeftAndRightWidths() // Needs to happen before setting up middle section
```
Comment (dafortin) [inline addon/components/frost-fixed-table.js]: Why do we need this now?

Comment (dafortin) [inline addon/components/frost-fixed-table.js]: Is this related to the selection?

File: addon/components/frost-fixed-table.js [hunk 6]
```diff
@@ -361,6 +453,28 @@ export default Component.extend({
...
+    },
+
+    _select ({isRangeSelect, isSpecificSelect, item}) {
+      const items = this.get('items')
+      const itemKey = this.get('itemKey')
+      const _itemComparator = this.get('_itemComparator')
+      const clonedSelectedItems = A(this.get('selectedItems').slice())
+      const _rangeState = this.get('_rangeState')
+
+      // Selects are proccessed in order of precedence: specific, range, basic
```
Comment (dafortin) [inline addon/components/frost-fixed-table.js]: I think this should be isolated in a util since it can be common.

File: addon/components/frost-table-row.js [hunk 7]
```diff
@@ -65,6 +65,11 @@ export default Component.extend({
 
   // == DOM Events ============================================================
 
+  click (event) {
+    event.stopPropagation()
+    this.$().find('.frost-table-row-selection').trigger('click')
```
Comment (dafortin) [inline addon/components/frost-table-row.js]: Can we simply trigger consumer action? (not sure here I'm simply wondering)

File: addon/templates/components/frost-fixed-table.hbs [hunk 8]
```diff
@@ -31,8 +31,10 @@
       hook=(concat hookPrefix '-header-left')
       hookQualifiers=hookQualifiers
       tagName='div'
+      isSelectable=_isSelectable
```
Comment (dafortin) [inline addon/templates/components/frost-fixed-table.hbs]: We can select the header? For sorting I'm expecting?

File: app/styles/ember-frost-table/_frost-fixed-table.scss [hunk 9]
```diff
@@ -48,6 +48,22 @@
     .ps-container > .ps-scrollbar-y-rail {
       display: none;
     }
+
+    // sass-lint:disable nesting-depth
+    .frost-table-row {
+      &.is-selected {
```
Comment (dafortin) [inline app/styles/ember-frost-table/_frost-fixed-table.scss]: Are we using `is-selected` in the list for selection? It will be nice to use the same class name.

File: addon/components/frost-fixed-table.js [hunk 10]
```diff
@@ -20,12 +21,29 @@ export default Component.extend({
   // == PropTypes =============================================================
 
   propTypes: {
-    // required:
+    // options:
```
Comment (AdamWard1995) [inline addon/components/frost-fixed-table.js]: Yes, consistency with frost-table

File: addon/components/frost-fixed-table.js (hunk 5 above)
Comment (AdamWard1995) [inline addon/components/frost-fixed-table.js]: Adding in the checkboxes column misaligns the columns BAD

File: addon/templates/components/frost-fixed-table.hbs (hunk 8 above)
Comment (AdamWard1995) [inline addon/templates/components/frost-fixed-table.hbs]: This was done in frost-table, for sorting was my guess as well but will look a little deeper to confirm

File: .travis.yml [hunk 11]
```diff
@@ -17,7 +17,6 @@ addons:
 
 cache:
   directories:
-  - node_modules
```
Comment (notmessenger) [inline .travis.yml]: @AdamWard1995 why?  This does not follow the practice of any of our other repos /cc @juwara0

Comment (juwara0) [inline .travis.yml]: @notmessenger - is correct. Please do not remove the caching.

Comment (job13er) [inline .travis.yml]: He's trying to figure out why something isn't working, and needed to clear the cache. It should be temporary.

Comment (AdamWard1995) [inline .travis.yml]: Going to add it back, just clearing it as locally everything is installing fine :)

Comment (juwara0) [inline .travis.yml]: That setting tells travis that it can cache the node_modules directories during builds. https://docs.travis-ci.com/user/caching/#Caching-directories-(Bundler%2C-dependencies). If you need to clear the cache during CI there is an option to do so on the build page under the drop-down in the upper right corner.

File: ember-cli-build.js [hunk 12]
```diff
@@ -24,8 +24,6 @@ module.exports = function (defaults) {
 
   if (app.env === 'test') {
```
Comment (notmessenger) [inline ember-cli-build.js]: The rest of this code

```
  if (app.env === 'test') {
    ;[
    ].forEach((path) => {
      app.import(path, {type: 'test'})
    })
  }
```

needs to be removed as well

Comment (AdamWard1995) [inline ember-cli-build.js]: OK

File: addon/components/frost-fixed-table.js (hunk 2 above)
Comment (AdamWard1995) [inline addon/components/frost-fixed-table.js]: huh? As far as I know there is no difference https://stackoverflow.com/questions/29885719/i-vs-i-in-a-javascript-for-loop. Just a habit of mine because post-increment creates and return a copy of the original value so as tiny as it may be, uses less memory. The pos + 1 is due to the fact that first element is selected with nth-child(1) where as the array starts at index 0

File: addon/components/frost-fixed-table.js (hunk 3 above)
Comment (AdamWard1995) [inline addon/components/frost-fixed-table.js]: yep

File: addon/components/frost-table-row.js (hunk 7 above)
Comment (AdamWard1995) [inline addon/components/frost-table-row.js]: So frost-table-row when you enable selection adds an frost-table-row-selection to the row. It is within the click event handler of frost-table-row-selection that the onSelect callback is call, thought it would be simplist to just trigger this event

File: addon/templates/components/frost-fixed-table.hbs (hunk 8 above)
Comment (AdamWard1995) [inline addon/templates/components/frost-fixed-table.hbs]: Not for sorting, this is how the header knows to add in a column in header that corresponds to the checkbox column

File: app/styles/ember-frost-table/_frost-fixed-table.scss (hunk 9 above)
Comment (AdamWard1995) [inline app/styles/ember-frost-table/_frost-fixed-table.scss]: Yep

File: addon/components/frost-table-row.js (hunk 7 above)
Comment (dafortin) [inline addon/components/frost-table-row.js]: Are we doing an action on click for the table on when we click on a button or something like that (similar to list)? If we are using a button to do an action based on selection then we should use a similar implementation to the list and I'm not seeing anything like this in the list.

**Comment by dafortin** [comment]
Looks good :) 👍
